### ISSUE TITLE
`invalid-assignment` with list item has different ranges compared to other type checkers

### ISSUE THREAD

**Issue by alexprengere** [astral-sh/ty#4450]
### Summary

If we look at those three versions of the same code, where the inline suppression comment is positioned at different places:

```python
# top
a: list[tuple[int, int]] = [  # type: ignore
    (1, 2),
    (3, 4),
    5,
]

# middle
a: list[tuple[int, int]] = [
    (1, 2),
    (3, 4),
    5,  # type: ignore
]

# bottom
a: list[tuple[int, int]] = [
    (1, 2),
    (3, 4),
    5,
]  # type: ignore
```
Ty has the following behavior
* top: 🟢 
* middle: 🔴 
* bottom: 🟢 

Mypy and pyright haves the following behavior:
* top: 🔴 
* middle: 🟢 
* bottom: 🔴 

I am not debating which one is wrong or right, but the inconsistency is a small paper cut (made worse by the fact that mypy has no mypy-specific suppression comment, like `mypy: ignore`).

**Comment by sharkdp** [comment]
Thank you for reporting this.

Adding this as a child issue to #995, which discusses a similar issue.

**Comment by MichaReiser** [comment]
The motivation for disallowing middle is documented here

https://github.com/astral-sh/ruff/blob/d165c3ca1236868f2f289244e107c04812fc1e58/crates/ty_python_semantic/resources/mdtest/suppressions/type_ignore.md?plain=1#L33-L35

ty's behavior is also closer to ruff, which I consider important for more users than cross type checker consistency.

**Comment by MichaReiser** [comment]
The issue here is also that ty uses a different diagnostic range than mypy and pyright. Fixing the range would fix the suppression behavior also:

**ty**

```python
a: list[tuple[int, int]] = [
                           ^
    (1, 2),
    (3, 4),
    5,
]
^
```

The range covers the entire list

**Mypy and Pyright**

```python
a: list[tuple[int, int]] = [
    (1, 2),
    (3, 4),
    5,
    ^
]
```

```python
a: list[tuple[int, int]] = [
    (1, 2),
    (3, 4),
    5,
    ^
]
```

Only highlight the problematic element. 

That's the reason why Mypy and Pyright accept the suppression on the element's line only, whereas ty accepts it on the list opener/closer.

The main difference between ty and mypy and Pyright's suppression behavior is that ty does not accept "in-the-middle" suppression, whereas mypy and Pyright allow any line between start and end.

Mypy also uses a dedicated rule `list-item`, pyright's `reportAssignmentType` rule is closer to ty's generic `invalid-assignment`.
